packages/application-shell/src/components/navbar/navbar.tsx

```diff
@@ -0,0 +1,667 @@
+          >
+            {hasSubmenu
+              ? props.menu.submenu.map(submenu => (
+                  <RestrictedMenuItem
+                    key={`${props.menu.key}-submenu-${submenu.key}`}
+                    keyOfMenuItem={submenu.key}
+                    featureToggle={submenu.featureToggle}
+                    permissions={submenu.permissions}
+                    actionRights={submenu.actionRights}
+                    dataFences={submenu.dataFences}
+                    menuVisibilities={props.menuVisibilities}
+                    namesOfMenuVisibilities={
+                      submenu.menuVisibility
+                        ? [submenu.menuVisibility]
+                        : undefined
+                    }
+                    disabledMenuItems={props.disabledMenuItems}
+                  >
+                    <li className={styles['sublist-item']}>
+                      <div className={styles.text}>
+                        <MenuItemLink
+                          linkTo={`/${props.projectKey}/${submenu.uriPath}`}
+                          exactMatch={true}
+                          useFullRedirectsForLinks={
+                            props.useFullRedirectsForLinks
+                          }
+                          onClick={props.onMenuItemClick}
+                        >
+                          {renderLabel(submenu.labelAllLocales)}
+                        </MenuItemLink>
+                      </div>
+                    </li>
+                  </RestrictedMenuItem>
+                ))
+              : null}
+          </MenuGroup>
+        </MenuItem>
+      </RestrictedMenuItem>
+    </>
+  );
+};
+ApplicationMenu.displayName = 'ApplicationMenu';
+
+type NavBarLayoutProps = {
+  children: React.ReactNode;
+};
+const NavBarLayout = React.forwardRef<HTMLElement, NavBarLayoutProps>(
+  (props, ref) => (
+    <nav
+      ref={ref}
+      className={styles['left-navigation']}
+      data-test="left-navigation"
+      data-testid="left-navigation"
+      data-track-component="Navigation"
+    >
+      {props.children}
+    </nav>
+  )
+);
+NavBarLayout.displayName = 'NavBarLayout';
+
+type NavbarProps = {
+  applicationLocale: string;
+  projectKey: string;
+  environment: TApplicationContext<{}>['environment'];
+  menuVisibilities: TApplicationContext<{}>['menuVisibilities'];
+  DEV_ONLY__loadNavbarMenuConfig?: () => Promise<TApplicationsMenu['navBar']>;
+  onMenuItemClick?: MenuItemLinkProps['onClick'];
+};
+const NavBar = (props: NavbarProps) => {
+  const {
+    navBarNode,
+    isMenuOpen,
+    isExpanderVisible,
+    activeItemIndex,
+    handleToggleItem,
+    handleToggleMenu,
+    shouldCloseMenuFly,
+    allApplicationNavbarMenu,
+  } = useNavbarStateManager({
+    environment: props.environment,
+    DEV_ONLY__loadNavbarMenuConfig: props.DEV_ONLY__loadNavbarMenuConfig,
+  });
+  const disabledMenuItems = props.environment.disabledMenuItems;
+  const menuVisibilities = useApplicationContext(
```

Comment:
It's not for this PR but those context accessors should often be `useMenuVisibilities` and build ontop of the application context but consumers don't care. We have sooo much code about this in the MC about languages and stuff. That simple wrappers could ease things and hide that detail.